### PR TITLE
Tap dance anyway

### PR DESCRIPTION
**Pull request by pvinis**
when using tap dance, we have the `regular` callback that is called on
the last tap. this commit adds an `anyway` callback that is called on
every tap, and a `reset` callback that is called on reset of the tap
dance taps.

use case: i want the following. one button is going to be the reset button, when i press it three times using tap dance. here is the need for anyway and reset. every time i tap it, one of the leds should go on.
```
enum {
  CT_RST = 0,
};

...

[BASE] = KEYMAP(
 TD(CT_RST) ,KC_NO  ,KC_NO

...

void dance_rst(qk_tap_dance_state_t *state) {
  if (state->count >= 3) {
    reset_keyboard();
    reset_tap_dance(state);
  }
}

void dance_rst_anyway(qk_tap_dance_state_t *state) {
  if (state->count == 1) {
    ergodox_right_led_3_on();
  } else if (state->count == 2) {
      ergodox_right_led_2_on();
  } else if (state->count == 3) {
    ergodox_right_led_1_on();
  }
}

void dance_rst_reset(void) {
  ergodox_led_all_off();
}

...

const qk_tap_dance_action_t tap_dance_actions[] = {
  [CT_RST] = ACTION_TAP_DANCE_FN_ANYWAY_RESET(dance_rst, dance_rst_anyway, dance_rst_reset)
};
```
when i tap the reset button once, the first led should go on, then if i dont press again, the reset is called, and my callback turns the leds off.
when i tap twice, the first led will go on with the first tap, the second will go on with the second tap, etc.
on the third tap, the third led turns on, and afterwards the keyboard will go to reset/flash mode.

some feedback from @algernon is needed.

these are the changes:
- add `qk_tap_dance_user_fn_reset_t` type. maybe we dont need it and we just use `qk_tap_dance_user_fn_t`? i just thought reset doesnt need the count or anything from the state.
- made a struct for the `fn` case, that holds the `regular` fn called on the final tap, the `anyway` fn called on every tap, and the `reset` fn called on tap dance reset.
- added `ACTION_TAP_DANCE_FN_ANYWAY`, `ACTION_TAP_DANCE_FN_RESET`, `ACTION_TAP_DANCE_FN_ANYWAY_RESET` macros for tap dance actions initialization. maybe we can just use `ACTION_TAP_DANCE_FN` with different number of args? variable args or overriding, or just always three args..?
- on tap dance reset, i call `action.fn.reset`.
- `process_tap_dance_action_anyway` is the way i call the `anyway` function. im not sure if where its called is correct or if it should be inside some check from the `if` below. also the commented call, i dont know if we need it. when does the `default` case get executed?

what do you think for each of those points?
thanks.

### CODE REVIEW

**Comment by algernon** [comment]
I have not looked at this closely yet, but will try to do so in the next couple of days. I like where this PR is going, but I'd suggest slightly different namings: `regular` is - to me, at least - does not convey the intent clearly enough. Something like `trigger` may do that better: that function gets called when the action needs to be done, when it is triggered.

I'm not sure about the `anyway` name, either, but at this time, I do not have a better suggestion yet.

As for the questions, in no particular order:

> when does the default case get executed?

That case gets executed when a non-tap-dance key is pressed. It's there so if you do a `TAP OTHER-KEY` sequence, the tap key will correctly trigger.

> add qk_tap_dance_user_fn_reset_t type. maybe we dont need it and we just use qk_tap_dance_user_fn_t? 

I'd use `qk_tap_dance_user_fn_t`, because the extra argument is not much of an overhead, and would allow one to do some action based on the state. For example, I'd like to have a key that when tapped once, registers one key, when tapped twice, registers another - and they unregister when the tap dance is reset. To know which key to unregister, I'll need the state.

Hope this helps! I'll come back and test the changes as soon as possible, and will try to provide more feedback. Thanks a lot for your work, by the way, this appears to pave the way to some of the things I wanted to do, too =)

**Comment by piotr-dobrogost** [comment]
I think names should describe when these functions are called. So something including phrases `on_tap`, `after_tap`, `tapped`, `each_tap` for one and including `final_tap`, `last_tap`, `dance_finished` for the other.

**Comment by algernon** [comment]
Now those are names I like even more! One note though, its the function that gets called that decides which is the last tap (unless it is interrupted), so `final_tap` and `last_tap` as callbacks doesn't make much sense. But the rest sound good!

**Comment by pvinis** [comment]
i still want to test if the `on_each_tap` function should be called where it is, or inside some if for checks.

are we happier with the names?

File: quantum/process_keycode/process_tap_dance.h [hunk 1]
```diff
@@ -31,18 +31,27 @@ typedef struct
...
-    .pair = { kc1, kc2 }                    \
+    .type = QK_TAP_DANCE_TYPE_PAIR, \
+    .pair = { kc1, kc2 }            \
   }
 
-#define ACTION_TAP_DANCE_FN(user_fn) { \
+#define ACTION_TAP_DANCE_FN(user_fn_) {  \
     .type = QK_TAP_DANCE_TYPE_FN, \
-    .fn = user_fn                 \
+    .fn = { NULL, user_fn, NULL } \
```
Comment (algernon) [inline quantum/process_keycode/process_tap_dance.h]: This looks wrong to me: `user_fn_` on the `#define` line, but `user_fn` here.

**Comment by algernon** [comment]
Apart from the one comment I made in-line for the code, this looks good to me, the naming too!

Can't wait to test it! :)

**Comment by algernon** [comment]
Took this for a drive today, and am liking it very much, the example given in the PR itself, too!

**Comment by piotr-dobrogost** [comment]
Out of curiosity; what would happen if one defined two or more sequences containing the same keys?
For example:
kc1, kc2, kc3, kc3, kc4, kc5 -> X
kc3, kc4, kc5 -> Y
kc3, kc3 -> Z

**Comment by pvinis** [comment]
with tap dance you tap multiple times of the same key. so what you suggest doesn't apply here.

**Comment by algernon** [comment]
Tap dance only deals with a single key, you can't have key sequences. The only sequence it deals with is how many times a single key was pressed within the tapping limit, without being interrupted.

What you're likely thinking of is the leader key. In that case, if you press `kc1 kc2 kc3 kc3 kc4 kc5`, you'll get X, because the sequence started at `kc1` and was not interrupted. That `kc3 kc3` is part of that sequence too, doesn't matter. A new sequence is not started until the previous one finished, timed out, or was interrupted.

**Comment by piotr-dobrogost** [comment]
Thanks guys for explanation.
I have another question then. What's going to happen if one defines given key both as a leader key and as a tap dance one? This question might seem odd but I keep wondering if anyone really knows how do different features interact? I mean by looking at source it's pretty hard to know which feature *overrides* which and if there aren't some that collide with each other.

Current approach is very imperative what makes interaction among different features rather hard to follow. I was wondering what if the model was changed to something pattern based? I mean we could have a fixed width window *observing* subsequent keyboard events (key up, key down) and a set of matching rules as regular expressions? This way we could free ourselves from keeping state as it would be handled automatically by regexp library.

**Comment by pvinis** [comment]
I haven't checked the way leader works but I'm sure you can make it so that tapping 'leader' once and then another key could do something, and tapping 'leader' twice would do something else. the second case is just tap dance. the first starts with tap dance, and it ends by timeout or by another key tapped, and then the leader logic takes over. it makes sense one you used one and the other separately. what you suggest sounds interesting though. maybe someone has to design and code the "automatic" part of what you said as a or first :p

**Comment by pvinis** [comment]
alright, i think this branch is good to go. unless, @algernon, you have any other concerns.

**Comment by algernon** [comment]
I'd prefer if `on_each_tap` would be called even in the depress case, not just when the tap key is pressed. I'd like to build something that requires knowledge of when a tap key is released too, and for that, `on_each_tap` being called in both cases is pretty much essential. I'd rather not introduce another callback for the depress case, that - I think - would complicate the code of what I want to build.

(I'll have to try actually building this idea, though, perhaps tomorrow, if all goes well.)

**Comment by algernon** [comment]
I took this for a ride today, and it's awesome. Already started to build stuff on top of it, as it makes a whole lot of things so much simpler!

Very nice work @pvinis, thank you!

**Comment by jackhumbert** [comment]
Awesome guys :)
